### ISSUE TITLE
Reentry is impossible to prevent when delegate calling

### ISSUE THREAD

**Issue by PaulRBerg** [nmushegian/wand#3]
The current implementation attempts to prevent reentry by checking that the `lock` is equal to the zero address at the time of entry in the `cast` function. But that can be circumvented in the target contract that is being delegate called. Here's an example:

```solidity
contract TargetReentry {
    address public root;
    address public auth;
    address public code;
    address public lock;

    function reentry() external returns (bool success, bytes memory response) {
        lock = address(0); // <-- problem here
        (success, response) = address(this).delegatecall(
            abi.encodeWithSelector(Wand.execute.cast, MALICIOUS_CONTRACT, "0x...")
        );
        if (!success) {
            if (response.length > 0) {
                assembly {
                    let returndata_size := mload(response)
                    revert(add(32, response), returndata_size)
                }
            } else {
                revert();
            }
        }
    }
}
```

For your convenience, the 4-byte signature for the `reentry` function is `0xb344d769`.

I'm not sure what purpose the reentry lock served in the first place. Maybe it would be worth it to remove it?

The delegate call in and of itself is like giving an external contract "god-like powers" over the original contract. I can't think of any attack that could be performed via reentry that couldn't be performed directly in the context of the delegate call.

**Comment by nmushegian** [comment]
You are right. This is an artifact of when this contract had the ability to have 'traps' (overrides) for certain casts, which were fetched from a different contract (like mudgen's 'diamond' contract), and the reentry lock was for that fetch. Later I separated that out but the lock remained.

**Comment by nmushegian** [comment]
https://ethereum-magicians.org/t/add-freeze-opcode/7231

**Comment by PaulRBerg** [comment]
>  'traps' (overrides) for certain casts

Interesting 🧐
